core/drivers/crypto/caam/caam_jr.c

```diff
@@ -0,0 +1,671 @@
+		jobctx->context = jobctx;
+	}
+
+	retstatus = do_jr_enqueue(jobctx, &jobctx->id);
+
+	if (retstatus != CAAM_NO_ERROR) {
+		JR_TRACE("enqueue job error 0x%08" PRIx32, retstatus);
+		return retstatus;
+	}
+
+	/*
+	 * If parameter job_id is defined, the job is asynchronous, so
+	 * returns with setting the job_id value
+	 */
+	if (job_id) {
+		*job_id = jobctx->id;
+		return CAAM_PENDING;
+	}
+
+#ifdef TIMEOUT_COMPLETION
+	/*
+	 * Job is synchronous wait until job complete or timeout
+	 */
+	while (!jobctx->completion && timeout--)
+		caam_jr_dequeue(jobctx->id, 100);
+
+	if (timeout <= 0) {
+		/* Job timeout, cancel it and return in error */
+		caam_jr_cancel(jobctx->id);
+		retstatus = CAAM_TIMEOUT;
+	} else {
+		if (JRSTA_SRC_GET(jobctx->status) != JRSTA_SRC(NONE))
+			retstatus = CAAM_JOB_STATUS;
+		else
+			retstatus = CAAM_NO_ERROR;
+	}
+#else
+	/*
+	 * Job is synchronous wait until job complete
+	 * Don't use a timeout because there is no HW timer and
+	 * so the timeout in not precise
+	 */
+	while (!jobctx->completion)
+		caam_jr_dequeue(jobctx->id, 100);
+
+	if (JRSTA_SRC_GET(jobctx->status) != JRSTA_SRC(NONE))
+		retstatus = CAAM_JOB_STATUS;
+	else
+		retstatus = CAAM_NO_ERROR;
+
+	/* Erase local callback function */
+	jobctx->callback = NULL;
+#endif
+
+	return retstatus;
+}
+
+enum CAAM_Status caam_jr_init(struct caam_jrcfg *jrcfg)
+{
+	enum CAAM_Status retstatus = CAAM_FAILURE;
+
+	JR_TRACE("Initialization");
+
+	/* Allocate the Job Ring resources */
+	retstatus = do_jr_alloc(&jr_privdata, jrcfg->nb_jobs);
+	if (retstatus != CAAM_NO_ERROR)
+		goto end_init;
+
+	jr_privdata->ctrladdr = jrcfg->base;
+	jr_privdata->jroffset = jrcfg->offset;
+
+	retstatus =
+		caam_hal_jr_setowner(jrcfg->base, jrcfg->offset, JROWN_ARM_S);
+	JR_TRACE("JR setowner returned 0x%" PRIx32, retstatus);
+
+	if (retstatus != CAAM_NO_ERROR)
+		goto end_init;
+
+	jr_privdata->baseaddr = jrcfg->base + jrcfg->offset;
+	retstatus = caam_hal_jr_reset(jr_privdata->baseaddr);
+	if (retstatus != CAAM_NO_ERROR)
+		goto end_init;
+
+	/*
+	 * Get the physical address of the Input/Output queue
+	 * The HW configuration is 64 bits registers regardless
+	 * the CAAM or CPU addressing mode.
+	 */
+	jr_privdata->paddr_inrings = virt_to_phys(jr_privdata->inrings);
+	jr_privdata->paddr_outrings = virt_to_phys(jr_privdata->outrings);
+	if (!jr_privdata->paddr_inrings || !jr_privdata->paddr_outrings) {
+		JR_TRACE("JR bad queue pointers");
+		retstatus = CAAM_FAILURE;
+		goto end_init;
+	}
+
+	caam_hal_jr_config(jr_privdata->baseaddr, jr_privdata->nb_jobs,
+			   jr_privdata->paddr_inrings,
+			   jr_privdata->paddr_outrings);
+
+	/*
+	 * Prepare the interrupt handler to secure the interrupt even
+	 * if the interrupt is not used
+	 */
+	jr_privdata->it_handler.it = jrcfg->it_num;
+	jr_privdata->it_handler.flags = ITRF_TRIGGER_LEVEL;
+	jr_privdata->it_handler.handler = caam_jr_irqhandler;
+	jr_privdata->it_handler.data = jr_privdata;
+
+#ifdef CFG_CRYPTO_DRIVER
```

Comment:
If the CFG_CRYPTO_DRIVER is not defined, it means that Job Ring is only used at boot time to configure RNG HW module (can be done only in Secure Environment) and then the Job Ring will be relaxed for the Non-Secure SW that will need the interrupt.
There are 2 reasons for the compilation flag;
- itr_add configure the GIC controller to set the IT number as secure interrupt (FIQ routed to secure). Hence Non-Secure SW can not access anymore this interrupt
- there is not itr_remove or itr_delete function to remove the secure interrupt configuration.

When CAAM Driver is used, the interrupt must be configured as secure to be not accessible in non-secure to ensure that its not caught by malware.

Reply:
Ok, I found the reference in an inline comment in `crypto_driver_init()`.
Maybe you could define a `CFG_NXP_CAAM_RUNTIME_JR=y` and use this instead of `CFG_CRYPTO_DRIVER` for all . This would make the code more easy to read. Their are places where it's not obvious why `#ifdef CFG_CRYPTO_DRIVER` condition instructions.

Reply:
ok